noir-projects/noir-contracts/contracts/fpc_contract/src/main.nr

```diff
@@ -56,38 +54,79 @@ contract FPC {
-    ///       and how much of it will be refunded back to the user. Since the protocol-calculated `tx_fee` is
-    ///       denominated in fee juice, and not in this FPC's AA, an equivalent value of AA is computed based
-    ///       on an exchange rate between AA and fee juice.
-    ///     - finalize the refund note with a value of `max_fee - tx_fee` for the user;
-    ///     - send the tx fee to the FPC in public.
+    ///     - Compute how much of the `max_fee` (denominated in AA) will be refunded back to the user. Since
+    ///       the protocol-calculated `tx_fee` is denominated in fee juice, and not in this FPC's AA, an equivalent
+    ///       value of AA is computed based on the exchange rate between AA and fee juice,
+    ///     - finalize the refund note with a value of `max_fee - tx_fee` for the user.
     ///
     /// Protocol-enshrined fee-payment phase:
     /// 4. The protocol deducts the protocol-calculated `tx_fee` (denominated in fee juice) from the `fee_payer`'s
     /// balance (which in this case is this FPC's balance), which is a special storage slot in a protocol-controlled
     /// "fee juice" contract.
     ///
     /// With this scheme a user has privately paid for the tx fee with an arbitrary AA (e.g. could be a stablecoin),
-    /// by paying this FPC. This FPC has in turn paid the protocol-mandated `tx_fee` (denominated in fee
-    /// juice).
+    /// by paying this FPC. This FPC has in turn paid the protocol-mandated `tx_fee` (denominated in fee juice).
     ///
     /// ***Note:***
     /// This flow allows us to pay for the tx with msg_sender's private balance of AA and hence msg_sender's identity
     /// is not revealed. We do, however, reveal:
-    /// - the `max_fee`;
-    /// - which FPC has been used to make the payment;
+    /// - the `max_fee`,
+    /// - which FPC has been used to make the payment,
     /// - the asset which was used to make the payment.
+    // docs:start:fee_entrypoint_private
     #[private]
     fn fee_entrypoint_private(max_fee: U128, nonce: Field) {
         // TODO(PR #8022): Once PublicImmutable performs only 1 merkle proof here, we'll save ~4k gates
-        let config = storage.config.read();
+        let accepted_asset = storage.config.read().accepted_asset;
 
-        Token::at(config.accepted_asset).setup_refund(context.msg_sender(), max_fee, nonce).call(
-            &mut context,
+        let user = context.msg_sender();
+        let token = Token::at(accepted_asset);
+
+        // TODO(#10805): Here we should check that `max_fee` converted to fee juice is enough to cover the tx
+        // fee juice/mana/gas limit. Currently the fee juice/AA exchange rate is fixed 1:1.
+
+        // Pull the max fee from the user's balance of the accepted asset to the public balance of this contract.
+        token.transfer_to_public(user, context.this_address(), max_fee, nonce).call(&mut context);
```

Comment:
It's not this PR's fault, but it's kind of ugly that this `nonce` has to pass-through this contract to get to the token contract. Do we have any ideas to improve that interface?

Reply:
The goal of the nonce there is to make the authwit valid only for the call with that specific nonce and prevent authwit re-use.

If we would like to get rid of it from the API we would have to achieve this some other way.

Given that authwits are so common maybe we could move the nonce from the function arguments to the context itself and then inject it into the authwit inner hash [here](https://github.com/AztecProtocol/aztec-packages/blob/dc7ec4853f6f44eec435b4e2c9ecb8e1e703861b/noir-projects/aztec-nr/authwit/src/auth.nr#L208)?

@iAmMichaelConnor WDYT?

Reply:
Yeah that might work. I've not paid much attention to authwits or capsules. I remember we wanted to refactor capsules, and maybe even that authwits would be _replaced_ by capsules (because capsules are a generalisation of authwits: injecting data into the function separately from params). Maybe we wait until the capsules refactor, if it's still planned?

Reply:
As I am looking at the code the usage of capsules was mostly removed (it's not even called from Noir anymore) so it seems like they have no future.

Tagging @nventuro as it would be nice to get a feedback on the nonce in context.